File: 백준/Gold/2493. 탑/탑.py
n = int(input())
tower = list(map(int, input().split()))
ans = [0] * n
stack = []

for i in range(n):
    while stack:
        if tower[stack[-1]] > tower[i]:
            ans[i] = stack[-1] + 1
            break
        else:
            stack.pop()
    stack.append(i)
 
print(*ans)

File: 백준/Gold/2493. 탑/탑.cc
#include <bits/stdc++.h>
using namespace std;

int main() {
	ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);

	int n; cin >> n;
	vector<int> tower(n);
	vector<int> ans(n, 0);
	stack<int> s;

	for (int i = 0; i < n; i++) cin >> tower[i];

	for (int i = 0; i < n; i++) {
		while (!s.empty()) {
			if (tower[s.top()] > tower[i]) {
				ans[i] = s.top() + 1;
				break;
			}
			s.pop();
		}
		s.push(i);
	}
	
	for (auto i : ans) cout << i << " ";

	return 0;
}